src/tpt_reports/report_generator.py

```diff
@@ -0,0 +1,423 @@
+"""Generate a TPT report using a passed data dictionary."""
+
+# Standard Python Libraries
+from datetime import datetime
+import os
+import secrets
+
+# Third-Party Libraries
+import numpy as np
+import pandas as pd
+from reportlab.lib import utils
+from reportlab.lib.colors import HexColor
+from reportlab.lib.pagesizes import letter
+from reportlab.lib.styles import ParagraphStyle
+from reportlab.lib.units import inch
+from reportlab.pdfbase import pdfmetrics
+from reportlab.pdfbase.ttfonts import TTFont
+from reportlab.platypus import (
+    Image,
+    KeepTogether,
+    PageBreak,
+    Paragraph,
+    Spacer,
+    Table,
+    TableStyle,
+)
+from reportlab.platypus.doctemplate import (
+    BaseDocTemplate,
+    NextPageTemplate,
+    PageTemplate,
+)
+from reportlab.platypus.frames import Frame
+
+BASE_DIR = os.path.abspath(os.path.dirname(__file__))
+TODAYS_DATE = datetime.today().strftime("%m/%d/%Y")
+
+# Set fonts to be used in the pdf
+pdfmetrics.registerFont(
+    TTFont("Franklin_Gothic_Book", f"{BASE_DIR}/fonts/FranklinGothicBook.ttf")
+)
+
+pdfmetrics.registerFont(
+    TTFont(
+        "Franklin_Gothic_Medium_Regular",
+        f"{BASE_DIR}/fonts/FranklinGothicMediumRegular.ttf",
+    )
+)
+
+# Set your page size
+defaultPageSize = letter
+PAGE_HEIGHT = defaultPageSize[1]
+PAGE_WIDTH = defaultPageSize[0]
+
+
+# Issue #23 - test generate_password()
+# TODO: Add unit tests for following logic and remove this comment.
+def generate_password(length):
+    """Generate a string for the password for the pdf."""
+    # Generate a random string of hexadecimal digits
+    password = secrets.token_hex(length // 2)
+
+    # Return the first `length` characters of the password
+    return password[:length]
```

Comment:
As this is written, this function will only return a password with a length of either `length` (when it is even) or `length-1` (when it is odd).  It seems like this is all you really need here:
```suggestion
    # Generate a random string of hexadecimal digits and
    # return the first `length` characters of the password
    return secrets.token_hex(length)[:length]
```

Reply:
That definitely makes sense to change to prevent issues down the road. Although if we only have the single return value in generate_password() that is third party library code it doesn't make sense to have the function at that point.

I made the suggested change in commit: 556d3c481b2f994349cb3949310ea3c974778a3f which cleans up the unnecessary function and updates usage downstream.

Reply:
Yessir, I like it! 🔥